module/zfs/vdev_trim.c

```diff
@@ -0,0 +1,1427 @@
+	vdev_xlate(vd, &logical_rs, &physical_rs);
+
+	IMPLY(vd->vdev_top == vd,
+	    logical_rs.rs_start == physical_rs.rs_start);
+	IMPLY(vd->vdev_top == vd,
+	    logical_rs.rs_end == physical_rs.rs_end);
+
+	/*
+	 * Only a manual trim will be traversing the vdev sequentially.
+	 * For an auto trim all valid ranges should be added.
+	 */
+	if (ta->trim_type == TRIM_TYPE_MANUAL) {
+
+		/* Only add segments that we have not visited yet */
+		if (physical_rs.rs_end <= vd->vdev_trim_last_offset)
+			return;
+
+		/* Pick up where we left off mid-range. */
+		if (vd->vdev_trim_last_offset > physical_rs.rs_start) {
+			zfs_dbgmsg("range write: vd %s changed (%llu, %llu) to "
+			    "(%llu, %llu)", vd->vdev_path,
+			    (u_longlong_t)physical_rs.rs_start,
+			    (u_longlong_t)physical_rs.rs_end,
+			    (u_longlong_t)vd->vdev_trim_last_offset,
+			    (u_longlong_t)physical_rs.rs_end);
+			ASSERT3U(physical_rs.rs_end, >,
+			    vd->vdev_trim_last_offset);
+			physical_rs.rs_start = vd->vdev_trim_last_offset;
+		}
+	}
+
+	ASSERT3U(physical_rs.rs_end, >=, physical_rs.rs_start);
+
+	/*
+	 * With raidz, it's possible that the logical range does not live on
+	 * this leaf vdev. We only add the physical range to this vdev's if it
+	 * has a length greater than 0.
+	 */
+	if (physical_rs.rs_end > physical_rs.rs_start) {
+		range_tree_add(ta->trim_tree, physical_rs.rs_start,
+		    physical_rs.rs_end - physical_rs.rs_start);
+	} else {
+		ASSERT3U(physical_rs.rs_end, ==, physical_rs.rs_start);
+	}
+}
+
+/*
+ * Each (manual) trim thread is responsible for trimming the unallocated
+ * space for each leaf vdev as described by its top-level ms_allocatable.
+ */
+static void
+vdev_trim_thread(void *arg)
+{
+	vdev_t *vd = arg;
+	spa_t *spa = vd->vdev_spa;
+	abd_t *deadbeef = NULL;
+	trim_args_t ta;
+	int error = 0;
+	uint64_t ms_count = 0;
+
+	/*
+	 * The VDEV_LEAF_ZAP_TRIM_* entries may have been updated by
+	 * vdev_trim().  Wait for the updated values to be reflected
+	 * in the zap in order to start with the requested settings.
+	 */
+	txg_wait_synced(spa_get_dsl(vd->vdev_spa), 0);
+
+	ASSERT(vdev_is_concrete(vd));
+	spa_config_enter(spa, SCL_CONFIG, FTAG, RW_READER);
+
+	vd->vdev_trim_last_offset = 0;
+	vd->vdev_trim_rate = 0;
+	vd->vdev_trim_partial = 0;
+	VERIFY0(vdev_trim_load(vd));
+
+	/*
+	 * When performing writes instead of discards the maximum extent
+	 * size needs to be capped at the maximum block size.
+	 */
+	uint64_t extent_bytes_max = zfs_trim_extent_bytes_max;
+	if (zfs_trim_write) {
+		extent_bytes_max = MIN(extent_bytes_max, SPA_MAXBLOCKSIZE);
+		deadbeef = vdev_trim_block_alloc(extent_bytes_max);
+	}
+
+	ta.trim_vdev = vd;
+	ta.trim_abd = deadbeef;
+	ta.trim_extent_bytes_max = extent_bytes_max;
+	ta.trim_extent_bytes_min = zfs_trim_extent_bytes_min;
+	ta.trim_tree = range_tree_create(NULL, NULL);
+	ta.trim_type = TRIM_TYPE_MANUAL;
+
+	for (uint64_t i = 0; !vd->vdev_detached &&
+	    i < vd->vdev_top->vdev_ms_count; i++) {
```

Comment:
The new manual trim algorithm, in which blocks are excluded from allocation via a "lock-out" mechanism (`vdev_trim_ms_mark()`), causes one interesting difference in behavior vs the exclusion algorithm used in the original Nexenta work:  it's much more likely to create new space maps for metaslabs which do not already have one during manual trim. Those who might use partial trim and/or large thinly-provisioned vdevs (SANs, etc.) may not like this behavior.  I tested a simple patch to iterate through the metaslabs in reverse during manual trim and it appears to have the desired effect of not causing lots of extra space maps to be created:
```
diff --git a/module/zfs/vdev_trim.c b/module/zfs/vdev_trim.c
index 75e39fa..69e8266 100644
--- a/module/zfs/vdev_trim.c
+++ b/module/zfs/vdev_trim.c
@@ -819,8 +819,8 @@ vdev_trim_thread(void *arg)
        ta.trim_tree = range_tree_create(NULL, NULL);
        ta.trim_type = TRIM_TYPE_MANUAL;
 
-       for (uint64_t i = 0; !vd->vdev_detached &&
-           i < vd->vdev_top->vdev_ms_count; i++) {
+       for (int64_t i = vd->vdev_top->vdev_ms_count - 1; !vd->vdev_detached &&
+           i >= 0; i--) {
                metaslab_t *msp = vd->vdev_top->vdev_ms[i];
 
                /*
```

Reply:
This may explain why my TRIM takes so long I am using partial trim and its a large thinly provisioned vdev.

Reply:
@bgly I'm not so sure, at worst partial TRIM performance would regress to be the same as a full TRIM.  For your specific setup I would be interesting to the the individual TRIM latency to see if that's slowing things down.  This depends heavily on how the backend storage/device handles the TRIM.  One way to get this time is by running `zpool iostat -w`.  For comparison, I've been seeing discard times of 1us - 4ms with my test hardware.

Reply:
IIRC, we saw similar behavior (touching all metaslabs) when using `zpool initialize`, but we determined that the impact was negligible.  But it might "look nicer" to start from the end, as @dweeezil suggested.

Reply:
Walking them in reverse order might also have the real benefit of front loading the most significant gains from TRIM which would be nice.  The downside is the current offset is stored on disk to resume from and it's expected to always increment.  If the general feeling is this is a worthwhile improvement, and we're OK with TRIM being different than initialize in this way, I can add it to the PR.

Reply:
> @bgly I'm not so sure, at worst partial TRIM performance would regress to be the same as a full TRIM. For your specific setup I would be interesting to the the individual TRIM latency to see if that's slowing things down. This depends heavily on how the backend storage/device handles the TRIM. One way to get this time is by running `zpool iostat -w`. For comparison, I've been seeing discard times of 1us - 4ms with my test hardware.

```
[root@bl-vsnap-100 ~]# zpool iostat -w                                                                                                                                                                                         [40/40]

rpool40dd706e0e594ef694f767328fbbabe9   total_wait     disk_wait    syncq_wait    asyncq_wait
latency                                 read  write   read  write   read  write   read  write  scrub   trim
-------------------------------------  -----  -----  -----  -----  -----  -----  -----  -----  -----  -----
1ns                                        0      0      0      0      0      0      0      0      0      0
3ns                                        0      0      0      0      0      0      0      0      0      0
7ns                                        0      0      0      0      0      0      0      0      0      0
15ns                                       0      0      0      0      0      0      0      0      0      0
31ns                                       0      0      0      0      0      0      0      0      0      0
63ns                                       0      0      0      0      0      0      0      0      0      0
127ns                                      0      0      0      0      0      0      0      0      0      0
255ns                                      0      0      0      0      0      0      0      0      0      0
511ns                                      0      0      0      0      0     11      0      0      0      0
1us                                        0      0      0      0      1    215      0     97      1      0
2us                                        0      0      0      0     26    102      0    495      0    109
4us                                        0      0      0      0     25     11      2     54      1    106
8us                                        0      0      0      0      4      4      1     38      0     39
16us                                       0      0      0      0      0      0      0     20      0     23
32us                                       0      0      0      0      0      0      0     12      0      6
65us                                       0      0      0      0      0      0      0     22      0      1
131us                                      1      5      1    231      0      0      0    121      0      1
262us                                      2    421      2  3.16K      0      0      0    407      0     27
524us                                      0  1.18K      0  1.34K      0      0      0    802      0     97
1ms                                        8  1.46K      8    185      0      0      0  1.23K      0    216
2ms                                       12  1.49K     12     46      0      0      0  1.13K      0  16.0K
4ms                                        4    370      5     22      0      0      0    217      0  1.51K
8ms                                        5     42      7      3      0      0      0     29      0     93
16ms                                       8     35     11      8      0      0      0     27      0     98
33ms                                      16      0     18      0      0      0      2      0      2      0
67ms                                       5      0      6      0      0      0      1      0      0      0
134ms                                      6      0      5      0      0      0      0      0      3      0
268ms                                      3      0      2      0      0      0      0      0      3      0
536ms                                      7      0      0      0      0      0      0      0      5      0
1s                                         0      0      0      0      0      0      0      0      0      0
2s                                         0      0      0      0      0      0      0      0      0      0
4s                                         0      0      0      0      0      0      0      0      0      0
8s                                         0      0      0      0      0      0      0      0      0      0
17s                                        0      0      0      0      0      0      0      0      0      0
34s                                        0      0      0      0      0      0      0      0      0      0
68s                                        0      0      0      0      0      0      0      0      0      0
137s                                       0      0      0      0      0      0      0      0      0      0
-----------------------------------------------------------------------------------------------------------
```

it seems like it is in line with what you are seeing